drafts/Asynchronous-programming-made-resilient-with-MPFT.adoc

```diff
@@ -0,0 +1,251 @@
+
+Even if the HTTP request fails, the request doesn't get retried. Why is this? Well, Fault Tolerance acts around method calls. Normally, if you annotate a method with `@Retry` and it throws an exception then it gets retried. However, when we do an HTTP request through this JAX-RS client API, if there's a problem it doesn't throw an exception. It can't throw an exception. It's going to do the request asynchronously, so when this method returns, it probably hasn't even made the request yet, let alone found out that it failed. If an exception does occur, it will be propagated  to the `CompletionStage` and can be handled there. But the result is, that this method will never throw an exception, even if the request fails and so the request will never be retried.
+
+*@Asynchronous to the rescue!*
+
+However, if we add the `@Asynchronous` annotation and the method returns a `CompletionStage`, then the Fault Tolerance logic gets applied when the `CompletionStage` completes, rather than when the method returns: 
+
+[source,java]
+----
+@Asynchronous
+@Retry
+public CompletionStage<String> clientDemo() {
+
+	...
+
+}
+----
+
+That means that when we call this method and it returns a `CompletionStage`, it's only when that `CompletionStage` completes that Fault Tolerance will look at the result and decide whether to Retry. So, if the request fails, the `CompletionStage` completes with an exception, Fault Tolerance decides that a Retry is needed and it calls the method again. As before, Fault Tolerance has intercepted the method call, so the `CompletionStage` returned to the caller is a different `CompletionStage` so that the caller doesn't get the result until all the Retries have been completed.
+
+So, to recap, to use Fault Tolerance with an asynchronous method you must:
+
+- **Return a `CompletionStage` from your method** - You can't do this with a `Future`, it must be a `CompletionStage`.
+- **Use the `@Asynchronous` annotation** - Without it, the method will never throw an exception, even if it fails.
+
+When you do these two things, all the other Fault Tolerance logic is applied when the `CompletionStage` completes, rather than when the method returns.
+
+You can also use https://download.eclipse.org/microprofile/microprofile-fault-tolerance-2.1/microprofile-fault-tolerance-spec.html#_interactions_with_other_fault_tolerance_annotations[other Fault Tolerance annotations with '@Asynchronous'] to make your asynchronous method resilient. See the <<FT-interactions, Interactions with other Fault Tolerance annotation>> section below for more detail about how the annotations interact.
+
+[#Running-methods-in-parallel]
+=== 2. Let’s go parallel!
+To run two methods in parallel, you can write methods which call two other services, annotate them both with the `@Asynchronous` annotation, then call them like this:
```

Comment:
To run two methods -> To run multiple methods
call two other services -> call other services
annotate them both -> annotate them

Reply:
Yep, I like this idea :)